### DISCUSSION TITLE
Exhaustively match all error variants

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#11828] [Ideas]
It would be great if the compiler could enforce exhaustive matches for error handling.

```v
fn default() ?int { return error('HELLO') }
fn opt_none() ?u64 { return none }

struct CustomError { msg string = 'hello' code int = 86 }
fn custom1() ?f32 { return IError(&CustomError{}) }

fn propagation() ? { custom1()? }

propagation() or {
    println('err.msg = $err.msg')
    match err {
        none { println('got `none`') }
        Error { println('got `Error`') }
        CustomError { println('got `CustomError`') }
        else { panic('unhandled error: $err') }           // <--- remove else
    }
}
```

This is a pretty difficult problem which can be discussed more when the language stabilizes.

**Comment by dumblob** [comment]
I agree - but I'd like to note, that this might be tricky to get right in case you want to explicitly handle only a certain subset of errors and for the rest just `return error_from_existing_error( err )` (i.e. not `panic()` as in the example above).
